server/src/main/java/org/opensearch/cluster/routing/allocation/decider/AwarenessAllocationDecider.java

```diff
@@ -209,7 +209,7 @@ private Decision underCapacity(ShardRouting shardRouting, RoutingNode node, Rout
             List<String> fullValues = forcedAwarenessAttributes.get(awarenessAttribute);
             if (fullValues != null) {
                 for (String fullValue : fullValues) {
-                    if (shardPerAttribute.containsKey(fullValue) == false) {
+                    if (nodesPerAttribute.containsKey(fullValue) == false) {
```

Comment:
Can we simplify the logic to reflect 
```
ceil(count(union(discovered_zone, force_zone))/total_shard_copies)
```

Reply:
> Could you please explain this change? AFAIK, the allocation is about shards, not nodes, as such usage of `shardPerAttribute` is about to be right (to me).

Yeah, thats true. But this implementation is showing incorrect attribute count in some cases when forced awareness is enabled for the domain which I mentioned in the issue. I have created a unit test which fails as well with incorrect awareness attribute count. 
           ```
 int numberOfAttributes = nodesPerAttribute.size();
            List<String> fullValues = forcedAwarenessAttributes.get(awarenessAttribute);
            if (fullValues != null) {
                for (String fullValue : fullValues) {
                    if (shardsPerAttribute.containsKey(fullValue) == false) {
                        numberOfAttributes++;
                    }
                }
            }
```

In code snippet, we are first calculating number of awareness attributes using discovered zones. Assume the domain has 3 zones but one zone (zone-c) is completely down, then nodesPerAttribute.size() returns 2. 
Assume forced awareness value is set to  "zone-a, zone-b, zone-c". So this returns three. ie fullValues =3. 

 Now assume a node came up in zone-c which had outage earlier. But no shards has been assigned to that node till now. So in this case shardsPerAttribute.size()=2 since there are no shards assigned to any node in zone-c but nodesPerAttribute.size()=3. Now when we go inside inner for loop, shardsPerAttribute won't have key for zone-c and it increases numberOfAttribute count to 4 which is incorrect.(ie zone-c has been counted twice)
So basically numAttributes = distinct(union(discovered_zone, forced_zone))

Reply:
> Could you please explain this change? AFAIK, the allocation is about shards, not nodes, as such usage of `shardPerAttribute` is about to be right (to me).

@reta  Yeah, thats true. But this implementation is showing incorrect attribute count in some cases when forced awareness is enabled for the domain which I mentioned in the issue. I have created a unit test which fails as well with incorrect awareness attribute count.

`

            int numberOfAttributes = nodesPerAttribute.size();
            List<String> fullValues = forcedAwarenessAttributes.get(awarenessAttribute);
            if (fullValues != null) {
                for (String fullValue : fullValues) {
                    if (shardPerAttribute.containsKey(fullValue) == false) {
                        numberOfAttributes++;
                    }
                }
            }
`

In code snippet, we are first calculating number of awareness attributes using discovered zones. Assume the domain has 3 zones but one zone (zone-c) is completely down, then nodesPerAttribute.size() returns 2. 
Assume forced awareness value is set to  "zone-a, zone-b, zone-c". So this returns three. ie fullValues =3. 

Now assume a node came up in zone-c which had outage earlier. But no shards has been assigned to that node till now. So in this case shardPerAttribute.size()=2 since there are no shards assigned to any node in zone-c but nodesPerAttribute.size()=3. Now when we go inside inner for loop, shardPerAttribute won't have key for zone-c and it increases numberOfAttribute count to 4 which is incorrect.(ie zone-c has been counted twice)
So basically numAttributes = distinct(union(discovered_zone, forced_zone))

Reply:
> Can we simplify the logic to reflect
> 
> ```
> ceil(count(union(discovered_zone, force_zone))/total_shard_copies)
> ```

Sure will try to make the logic simpler.

Reply:
I think it makes sense now, thank @anshu1106 !